File: Data Structures in JAVA/BinarySearchTree.java
package DataStructure;

import java.util.Scanner;

enum NodeType{
    UNKNOWN, NO_CHILD, ONE_CHILD,DOUBLE_CHILDREN
}
enum OrderType{
    PRE_ORDER,INTERIM_ORDER,POST_ORDER
}
public class BinarySearchTree {
    static class TreeNode{
        //建立节点
        protected int key;
        protected String value;
        protected TreeNode left;
        protected TreeNode right;
        public TreeNode(int key, String value){
            this.key = key;
            this.value = value;
            this.left = null;
            this.right = null;
        }
    }

    private TreeNode root;//记录根节点
    private int count;

    public void insert(int key,String value){
        /*插入函数*/
        if(root==null){
            root = new TreeNode(key,value);
            return;
        }
        TreeNode current = root;
        TreeNode parent = null;
        while(true){
            parent = current;
            if(key<parent.key){
                current = parent.left;
                if(current == null){
                    parent.left = new TreeNode(key,value);
                    return;
                }
            }
            else if(key>parent.key){
                current = parent.right;
                if(current==null){
                    parent.right = new TreeNode(key,value);
                    return;
                }
            }
            else{
                return;//BST doesn't allow nodes with same key
            }
        }
    }

    public TreeNode get(int key){
        /*查询方法*/
        TreeNode current = root;
        while(current!=null && current.key!=key){
            if(key<current.key){
                current = current.left;
            }
            else{
                current = current.right;
            }
        }
        return current == null ? null:current;
    }

    public boolean delete(int key){
        TreeNode parent = root;
        TreeNode current = root;
        boolean isLeftChild = false;
        while(current!=null && current.key!=key){
            //先找到该节点
            parent = current;
            if(key<current.key){
                //在左节点
                isLeftChild=true;
                current =current.left;
            }
            else{
                //在右节点
                isLeftChild=false;
                current = current.right;
            }
        }
        if(current==null){
            return false;//没有该节点
        }
        // case 1: if node to be deleted has no children
        NodeType nodeType = NodeType.UNKNOWN;
        if(current.left==null && current.right==null){
            nodeType = NodeType.NO_CHILD;
        }
        // case 2: if node to be deleted has onley one child
        else if((current.right==null)^(current.left==null)){
            // ^ 表示相异
            nodeType = NodeType.ONE_CHILD;
        }
        // case 3: if node to be deleted has double children
        else{
            nodeType = NodeType.DOUBLE_CHILDREN;
        }
        switch(nodeType){
            case NO_CHILD:
                if(current==root){
                    root = null;//只有一个节点
                }
                else if(isLeftChild){
                    //如果是左节点
                    parent.left =null;
                }else if(isLeftChild==false){
                    parent.right =null;
                }
                break;
            case ONE_CHILD:
                if(current==root){
                    if(isLeftChild){
                        root = current.left;
                    }
                    else{
                        root = current.right;
                    }
                }
                else if(isLeftChild){
                    parent.left = current.left;
                }
                else{
                    parent.right = current.right;
                }
                break;
            case DOUBLE_CHILDREN:
                TreeNode successor = getSuccessor(current);
                if(current == root){
                    root = successor;
                }
                else if(isLeftChild){
                    parent.left = successor;
                }else{
                    parent.right = successor;
                }
                successor.left = current.left;
                break;
            default:
                //以上都没选到
                break;
        }
        return true;
    }

    private TreeNode getSuccessor(TreeNode node){
        /* 找到右子树最小值，则需要在右子树下一直往左找，找左子节点leaf node
        * */
        TreeNode successor = null;//最后替代的节点
        TreeNode successorParent = null;//遍历记录父节点
        TreeNode current = node.right;//进入右子树
        while(current!=null){
            successorParent=successor;
            successor=current;
            current=current.left;
        }
        if(successor!=node.right){
            successorParent.left=successor.right;
            successor.right = node.right;
        }
        return successor;
    }

    private void traversal(OrderType _type){
        /*这里开始选择何种遍历方式进行遍历*/
        count =0;
        TreeNode current = root;
        switch(_type){
            case PRE_ORDER:
                System.out.println("preOrderTraversal below:");
                preOrderTraversal(current);
                break;
            case INTERIM_ORDER:
                System.out.println("interimOrderTraversal below:");
                interimOrderTraversal(current);
                break;
            case POST_ORDER:
                System.out.println("postOrderTraversal below:");
                postOrderTraversal(current);
                break;
        }
    }

    private void preOrderTraversal(TreeNode current){
        if(current==null){
            return;
        }
        count+=1;
        System.out.println("No "+count+" of Traversal is: key: "+current.key+" value: "+current.value);
        preOrderTraversal(current.left);
        preOrderTraversal(current.right);
    }

    private void interimOrderTraversal(TreeNode current){
        if(current==null){
            return;
        }
        count+=1;
        interimOrderTraversal(current.left);
        System.out.println("No "+count+" of Traversal is: key: "+current.key+" value: "+current.value);
        interimOrderTraversal(current.right);
    }

    private void postOrderTraversal(TreeNode current){
        if(current==null){
            return;
        }
        count+=1;
        postOrderTraversal(current.left);
        postOrderTraversal(current.right);
        System.out.println("No "+count+" of Traversal is: key: "+current.key+" value: "+current.value);
    }


    public static void main(String[] args){
        Scanner obj = new Scanner(System.in);
        System.out.println("pleae enter the node with format: int key,String value \\n:");
        BinarySearchTree bst = new BinarySearchTree();
        while(obj.hasNextLine()){
            String strInput = obj.nextLine();
            String[] listInput = strInput.split(",");
            int key = Integer.parseInt(listInput[0]);
            String value = listInput[1];
            bst.insert(key, value);
        }
        bst.traversal(OrderType.PRE_ORDER);
        bst.delete(4);
        bst.delete(28);
        bst.traversal(OrderType.PRE_ORDER);
    }
}

File: Data Structures in Python/BinarySearchTree.py

class TreeNode:
    def __init__(self,key):
        self.left = None
        self.right= None
        self.val  = key

class BinarySearchTree:
    def __init__(self):
        self.root = None # 根节点
        self.count = -1 # 遍历计算顺序序号

    def get(self,key):
        """
        查询功能
        """
        cur = self.root

        # recursive loop
        while cur!= None and cur.val !=key:
            if key<cur.val:
                cur = cur.left
            elif key > cur.val:
                cur = cur.right

        # Verdict on whether cur exists
        if cur is not None:
            return cur
        else:
            return None

    def insert(self,key):
        """
        whether you can insert a new kye into the tree:return
        """
        if self.root is None:
            self.root = TreeNode(key)
            return
        else:
            cur = self.root
            parent = None
            while True:
                parent = cur
                if key < parent.val:
                    cur = parent.left
                    if cur is None:
                        parent.left = TreeNode(key)
                        return
                elif key>parent.val:
                    cur = parent.right
                    if cur is None:
                        parent.right = TreeNode(key)
                        return
                else:
                    return # BST 不允许相同的值存在

    def delete(self,key):
        """
        三种情况：
            1.删除Leaf
            2.删除只有一个Child的节点
            3.删除有两个Child的节点
        """
        parent = self.root
        cur = self.root
        isLeftChild = False # 判断要删除节点与父节点的关系
        while cur != None and cur.val != key:
            parent = cur
            if key < cur.val:
                isLeftChild = True
                cur = cur.left
            elif key > cur.val:
                isLeftChild = False
                cur = cur.right

        if cur is None:#如果不存在该key
            return False

        # case 1: 没有子节点的Leaf
        if cur.left is None and cur.right is None:
            if cur == self.root:#如果是头节点，且只有一个元素
                self.root = None
            elif isLeftChild == True:
                parent.left = None
            else:
                parent.right = None

        # case 2: 如果节点只有一个子节点
        elif cur.right is None: #只有左子树，没有右子树
            if cur == self.root:
                self.root = cur.left
            elif isLeftChild == True:
                parent.left = cur.left
            elif isLeftChild == False:
                parent.right = cur.left

        elif cur.left is None:  # 只有右子树，没有左子树
            if cur == self.root:
                self.root = cur.right
            elif isLeftChild == True:
                parent.left = cur.right
            elif isLeftChild == False:
                parent.right = cur.right

        # case 3： 如果节点有两个子树 cur.left is not None and cur.right is not None
        else:
            #这里我们找右子树最小节点来替代要删除的点
            successor = self.getSuccessor(cur) #找右子树最小的节点，successor即胜出的最小节点，也把关系都清理了
            if cur == self.root:
                self.root = successor
            elif isLeftChild == True:
                parent.left = successor
            else:
                parent.right = successor

        return True


    def getSuccessor(self,node:TreeNode):
        """:arg
        找右子树最小值
        """
        successor = None #最后替代的节点
        successor_parent = None
        cur = node.right
        while cur!=None:
            successor_parent = successor
            successor = cur
            cur = cur.left

        if successor!=node.right:
            successor_parent.left = successor.right
            successor.right = node.right

        successor.left = node.left

        return successor

    def traversal(self,type=1):
        self.count = -1
        cur = self.root
        if type ==1:#前序遍历
            self.pre_orderTraversal(cur)
        elif type ==2:#中序遍历
            self.interim_orderTraversal(cur)
        elif type ==3:#后序遍历
            self.post_orderTraversal(cur)

    def pre_orderTraversal(self,cur):
        """
        前序遍历
        """
        if cur is None:
            return
        else:
            self.count +=1
            print("No {a} of Traversal is: {b}\n".format(a=self.count,b=cur.val))
            self.pre_orderTraversal(cur.left)
            self.pre_orderTraversal(cur.right)


    def interim_orderTraversal(self,cur):
        """
        中序遍历
        """
        if cur is None:
            return
        else:
            self.count +=1
            self.pre_orderTraversal(cur.left)
            print("No {a} of Traversal is: {b}\n".format(a=self.count,b=cur.val))
            self.pre_orderTraversal(cur.right)

    def post_orderTraversal(self,cur):
        """
        后序遍历
        """
        if cur is None:
            return
        else:
            self.count +=1
            self.pre_orderTraversal(cur.left)
            self.pre_orderTraversal(cur.right)
            print("No {a} of Traversal is: {b}\n".format(a=self.count, b=cur.val))


if __name__ =="__main__":
    list_insert = [2,1,5,4,-3,43,22,5,8,21,16]
    bst = BinarySearchTree()
    for each in list_insert:
        bst.insert(each)

    print(bst.get(21).val)
    bst.traversal()
    bst.delete(16)
    bst.traversal(type=3)